presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -114,11 +108,29 @@ public void removeListener(MemoryPoolListener listener)
         checkArgument(bytes >= 0, "bytes is negative");
 
         ListenableFuture<?> result;
-        synchronized (this) {
-            if (bytes != 0) {
-                queryMemoryReservations.merge(queryId, bytes, Long::sum);
-                updateTaggedMemoryAllocations(queryId, allocationTag, bytes);
+        if (bytes != 0) {
+            boolean done = false;
+            while (!done) {
```

Comment:
I think using done vs break is less clear?
```
            while (true) {
                if (queryMemoryReservations.computeIfPresent(queryId, (k, v) -> v + bytes) != null) {
                    // queryMemoryReservations is updated,
                    // use flag putQueryIdAbsent = false to prevent adding new queryId to taggedMemoryAllocations
                    // in race condition when queryId has been removed
                    updateTaggedMemoryAllocations(queryId, allocationTag, bytes, false);
                    break;
                }
                synchronized (this) {
                    if (queryMemoryReservations.putIfAbsent(queryId, bytes) == null) {
                        // queryId is successfully put into queryMemoryReservations,
                        // safe to update taggedMemoryAllocations
                        updateTaggedMemoryAllocations(queryId, allocationTag, bytes, true);
                        break;
                    }
                }
            }
```